### PR TITLE
Refactor got handling infection stage and progess of the black goo disease

### PR DESCRIPTION
**Pull request by Huffie56**
# About the pull request

1- reduce the number of stages from 5 to 3. each stage take 6 min or 360s

2-rework the logic that handle the disease progress

A-Set stage_prob to zero to prevent random stage increase.

B-add a stage_level variable that when is at-least 360 make an increase of the disease stage...

C-add an infection_rate variable that handle how fast the stage_level increase passively.
( stage_level = stage_level +infection_rate )
i also made it so that infection-rate is quadruple when the host is dead.

D-I also added that the goo messages (you feel warm ....) give a boost to stage_level to add some randomness to the infection progress.

To do list:
i need to figure out how to increase stage level when someone get hit by a zombie after he is infected...
no idea if i can manage that so i am just going to put what i made up for review...

# Explain why it's good for the game
those change are to make infection evolution less random and way more smooth progress with the addition of some random boost with the good message since it's something that could represent a natural boost of the infection progress in the host.
# Testing Photographs and Procedure
<details>
<summary>Screenshots & Videos</summary>

Put screenshots and videos here with an empty line between the screenshots and the `<details>` tags.

</details>

# Changelog
:cl:
balance: reduce the number of stages from 5 to 3. each stage take 6 min or 360s.
balance: rework the logic that handle the disease progress
balance: A-Set stage_prob to zero to prevent random stage increase.
balance: B-add a stage_level variable that when is at-least 360 make an increase of the disease stage...
balance: C-add an infection_rate variable that handle how fast the stage_level increase passively.
balance: D-infection_rate is quadruple when the host is dead.
balance: E-I also added that the goo messages (you feel warm ....) give a boost to stage_level to add some randomness to the infection progress.
/:cl:

### CODE REVIEW

File: code/datums/diseases/black_goo.dm [hunk 1]
```diff
@@ -1,100 +1,106 @@
...
 			if(goo_message_cooldown < world.time)
-				if (prob(3))
-					to_chat(affected_mob, SPAN_DANGER("You cough up some black fluid..."))
-					goo_message_cooldown = world.time + 100
-				else if (prob(6))
+				if (prob(5))
 					to_chat(affected_mob, SPAN_DANGER("Your throat is really dry..."))
-					goo_message_cooldown = world.time + 100
-				else if (prob(9))
+					stage_level = stage_level + 10
```
Comment (Zonespace27) [inline code/datums/diseases/black_goo.dm:76]: use +=

File: code/datums/diseases/black_goo.dm [hunk 2]
```diff
@@ -1,100 +1,106 @@
...
-				stage_counter = stage
-			if(goo_message_cooldown < world.time)
-				if (prob(3)) to_chat(affected_mob, SPAN_DANGER("Your throat is really dry..."))
-				else if (prob(6)) to_chat(affected_mob, SPAN_DANGER("You feel really warm..."))
-				else if (prob(2)) H.vomit_on_floor()
-				goo_message_cooldown = world.time + 100
-		if(3)
-			if(H.stat == DEAD && stage_counter != stage)
-				to_chat(H, SPAN_CENTERBOLD("Your zombie infection is now at Stage Three! Zombie transformation begins at Stage Four, which will be soon."))
+				to_chat(H, SPAN_CENTERBOLD("Your zombie infection is now at Stage Two! Zombie transformation begins at Stage Three."))
```
Comment (Zonespace27) [inline code/datums/diseases/black_goo.dm:71]: stage numbers aren't proper nouns.
Centering this seems 😕

File: code/datums/diseases/black_goo.dm [hunk 3]
```diff
@@ -1,100 +1,106 @@
...
+
+		//stage 3 : Lusus "Freak"
+		if(3)
+			//check if your already a zombie just return to avoid weird stuff... if for some weird reason first filter deoesn't work...
+			if(H.species.name == SPECIES_ZOMBIE || zombie_transforming == TRUE)
+				return
+
 			if(H.stat == DEAD && stage_counter != stage)
-				to_chat(H, SPAN_CENTERBOLD("Your zombie infection is now at Stage Four! Your transformation will happen any moment now."))
+				to_chat(H, SPAN_CENTERBOLD("Your zombie infection is now at Stage Three! Zombie transformation begin!"))
```
Comment (Zonespace27) [inline code/datums/diseases/black_goo.dm:98]: stages aren't proper nouns, please grammar check this.

File: code/datums/diseases/black_goo.dm [hunk 4]
```diff
@@ -1,100 +1,106 @@
...
+					stage_level = stage_level + 15
+				else if(prob(3))
+					to_chat(affected_mob, SPAN_DANGER("You can't trust them..."))
+					stage_level = stage_level + 4
+				goo_message_cooldown = world.time + 100
+
+		//stage 3 : Lusus "Freak"
+		if(3)
+			//check if your already a zombie just return to avoid weird stuff... if for some weird reason first filter deoesn't work...
+			if(H.species.name == SPECIES_ZOMBIE || zombie_transforming == TRUE)
```
Comment (Zonespace27) [inline code/datums/diseases/black_goo.dm:94]: == TRUE not needed

File: code/datums/diseases/black_goo.dm [hunk 5]
```diff
@@ -1,100 +1,106 @@
...
+					to_chat(affected_mob, SPAN_DANGER("You cough up some black fluid..."))
+					stage_level = stage_level + 20
+				else if (prob(2))
 					H.vomit_on_floor()
-		if(4)
+					stage_level = stage_level + 15
+				else if(prob(3))
+					to_chat(affected_mob, SPAN_DANGER("You can't trust them..."))
+					stage_level = stage_level + 4
+				goo_message_cooldown = world.time + 100
```
Comment (Zonespace27) [inline code/datums/diseases/black_goo.dm:89]: use cooldown macros

File: code/datums/diseases/black_goo.dm [hunk 6]
```diff
@@ -1,100 +1,106 @@
...
+	survive_mob_death = TRUE //We want the dead to turn into zombie.
+	longevity = 500 //the virus tend to die before the dead is turn into zombie this should fix it.
 	var/zombie_transforming = 0 //whether we're currently transforming the host into a zombie.
 	var/goo_message_cooldown = 0 //to make sure we don't spam messages too often.
 	var/stage_counter = 0 // tells a dead infectee their stage, so they can know when-abouts they'll revive
 
+	// set of new var to handle stage progress and
+	var/stage_level = 0 // we start at zero when dead everything go twice as fast each stage take 6 min so it need 360 and if your dead it require 180 so 3 min
+	var/infection_rate = 1
+	stage_prob = 0//no randomness
```
Comment (Zonespace27) [inline code/datums/diseases/black_goo.dm:25]: parent vars above new ones

File: code/datums/diseases/black_goo.dm [hunk 7]
```diff
@@ -1,100 +1,106 @@
...
+
+	survive_mob_death = TRUE //We want the dead to turn into zombie.
+	longevity = 500 //the virus tend to die before the dead is turn into zombie this should fix it.
 	var/zombie_transforming = 0 //whether we're currently transforming the host into a zombie.
 	var/goo_message_cooldown = 0 //to make sure we don't spam messages too often.
 	var/stage_counter = 0 // tells a dead infectee their stage, so they can know when-abouts they'll revive
 
+	// set of new var to handle stage progress and
+	var/stage_level = 0 // we start at zero when dead everything go twice as fast each stage take 6 min so it need 360 and if your dead it require 180 so 3 min
+	var/infection_rate = 1
```
Comment (Zonespace27) [inline code/datums/diseases/black_goo.dm:24]: convert to code doc

File: code/datums/diseases/black_goo.dm [hunk 8]
```diff
@@ -1,100 +1,106 @@
...
+	// here we add the new infection rate to the stage level.
+	stage_level = stage_level + infection_rate
+
+	// we want to check if we have reach enough stage level to gain a stage 3 stage of 6 min if you get it once.
+	if(stage_level >= 360)
+		stage ++
+		stage_level = stage_level - 360
+
 	switch(stage)
+		//stage 1 : Viatiate "afflicted"
```
[the remark this answers is not in the document]
Comment (Huffie56) [reply code/datums/diseases/black_goo.dm:53]: that's the lore name i found on a wiki for this stage.
i removed it with the other stages name.

File: code/datums/diseases/black_goo.dm [hunk 9]
```diff
@@ -1,100 +1,106 @@
...
+				if(prob(5))
+					to_chat(affected_mob, SPAN_DANGER("You feel warm..."))
 					goo_message_cooldown = world.time + 100
+					stage_level = stage_level + 4
+				else if(prob(3))
+					to_chat(affected_mob, SPAN_DANGER("You can't trust them..."))
+					goo_message_cooldown = world.time + 100
+					stage_level = stage_level + 2
+
+		//stage 2 : Ague "Febrile"
```
[the remark this answers is not in the document]
Comment (Huffie56) [reply code/datums/diseases/black_goo.dm:68]: removed

File: code/datums/diseases/black_goo.dm (hunk 1 above)
Comment (Huffie56) [reply code/datums/diseases/black_goo.dm:76]: done also use -= for one.

File: code/datums/diseases/black_goo.dm (hunk 7 above)
Comment (Huffie56) [reply code/datums/diseases/black_goo.dm:24]: done

File: code/datums/diseases/black_goo.dm [hunk 10]
```diff
@@ -1,100 +1,106 @@
...
 /datum/disease/black_goo/stage_act()
 	..()
 	if(!ishuman(affected_mob)) return
 	var/mob/living/carbon/human/H = affected_mob
 
-	if(age > 1.5*stage_minimum_age) stage_prob = 100 //if it takes too long we force a stage increase
-	else stage_prob = initial(stage_prob)
-	if(H.stat == DEAD) stage_minimum_age = 75 //the virus progress faster when the host is dead.
+	// check if your already a zombie or in the process of being transform into one...
+	if(H.species.name == SPECIES_ZOMBIE || zombie_transforming == TRUE)
```
[the remark this answers is not in the document]
Comment (Huffie56) [reply code/datums/diseases/black_goo.dm:33]: i replace it by iszombie(H) and == true is removed.

File: code/datums/diseases/black_goo.dm (hunk 4 above)
Comment (Huffie56) [reply code/datums/diseases/black_goo.dm:94]: i replace it by iszombie(H) and == true is removed.

File: code/datums/diseases/black_goo.dm (hunk 5 above)
Comment (Huffie56) [reply code/datums/diseases/black_goo.dm]: done was a pain to understand the logic but i manage to understood it thanks to fira.

File: code/datums/diseases/black_goo.dm (hunk 2 above)
Comment (Huffie56) [reply code/datums/diseases/black_goo.dm]: change Stage to stage and did the same for everything that had maj and shouldn't

File: code/datums/diseases/black_goo.dm (hunk 3 above)
Comment (Huffie56) [reply code/datums/diseases/black_goo.dm]: change Stage to stage and did the same for everything that had maj and shouldn't

File: code/datums/diseases/black_goo.dm [hunk 11]
```diff
@@ -1,100 +1,120 @@
...
+//new variables to handle infection progression inside a stage.
+
+	/// variable that contain accumulated virus progression for an host.
+	var/stage_level = 0
+	/// variable that handle passive increase of the virus of an host.
+	var/infection_rate = 1
+
+//cooldown
+
+	var/message_cooldown_time = 100
```
Comment (Zonespace27) [inline code/datums/diseases/black_goo.dm:35]: code doc this, use time macros

File: code/datums/diseases/black_goo.dm [hunk 12]
```diff
@@ -1,100 +1,120 @@
...
+	// check if he isn't dead
+	if(H.stat != DEAD)
+		infection_rate = 1
+
+	// here we add the new infection rate to the stage level.
+	stage_level += infection_rate
+
+	// we want to check if we have reach enough stage level to gain a stage 3 stage of 6 min if you get it once.
+	if(stage_level >= 360)
+		stage ++
```
Comment (Zonespace27) [inline code/datums/diseases/black_goo.dm:61]:
```suggestion
		stage++
```

File: code/datums/diseases/black_goo.dm [hunk 13]
```diff
@@ -1,100 +1,120 @@
...
+
+	// check if he isn't dead
+	if(H.stat != DEAD)
+		infection_rate = 1
+
+	// here we add the new infection rate to the stage level.
+	stage_level += infection_rate
+
+	// we want to check if we have reach enough stage level to gain a stage 3 stage of 6 min if you get it once.
+	if(stage_level >= 360)
```
Comment (Zonespace27) [inline code/datums/diseases/black_goo.dm:60]: make 360 a variable or define

File: code/datums/diseases/black_goo.dm [hunk 14]
```diff
@@ -1,100 +1,120 @@
...
+				stage_level += 5
+			else if (prob(2))
+				to_chat(affected_mob, SPAN_DANGER("You cough up some black fluid..."))
+				stage_level += 20
+			else if (prob(2))
+				H.vomit_on_floor()
+				stage_level += 15
+			else if(prob(3))
+				to_chat(affected_mob, SPAN_DANGER("You can't trust them..."))
+				stage_level += 4
```
Comment (Zonespace27) [inline code/datums/diseases/black_goo.dm:104]: these prob() elseifs could be turned into a rand() with a switch statement

File: code/datums/diseases/black_goo.dm [hunk 15]
```diff
@@ -1,100 +1,120 @@
...
-					goo_message_cooldown = world.time + 100
+
+			if (!COOLDOWN_FINISHED(src, goo_message_cooldown))
+				return
+			COOLDOWN_START(src, goo_message_cooldown, message_cooldown_time)
+
+			if(prob(5))
+				to_chat(affected_mob, SPAN_DANGER("You feel warm..."))
+				stage_level += 4
+			else if(prob(3))
```
Comment (Zonespace27) [inline code/datums/diseases/black_goo.dm:77]: same here

**Review by Zonespace27 — CHANGES_REQUESTED** [review]

File: code/datums/diseases/black_goo.dm (hunk 13 above)
Comment (Huffie56) [reply code/datums/diseases/black_goo.dm]: made it a variable

File: code/datums/diseases/black_goo.dm (hunk 15 above)
Comment (Huffie56) [reply code/datums/diseases/black_goo.dm]: done seem way more solid for here.

File: code/datums/diseases/black_goo.dm (hunk 12 above)
Comment (Huffie56) [reply code/datums/diseases/black_goo.dm]: done

File: code/datums/diseases/black_goo.dm (hunk 11 above)
Comment (Huffie56) [reply code/datums/diseases/black_goo.dm]: found it what it mean so i replace it

File: code/datums/diseases/black_goo.dm [hunk 16]
```diff
@@ -1,100 +1,130 @@
...
-	var/goo_message_cooldown = 0 //to make sure we don't spam messages too often.
-	var/stage_counter = 0 // tells a dead infectee their stage, so they can know when-abouts they'll revive
+	survive_mob_death = TRUE //We want the dead to turn into zombie.
+	longevity = 500 //the virus tend to die before the dead is turn into zombie this should fix it.
+	stage_prob = 0//no randomness
+
+	/// whether we're currently transforming the host into a zombie.
+	var/zombie_transforming = 0
+	/// to make sure we don't spam messages too often.
+	//var/goo_message_cooldown = 0
```
Comment (Zonespace27) [inline code/datums/diseases/black_goo.dm:22]:
```suggestion
```

**Review by Zonespace27 — APPROVED** [review]

**Pull request merged**